### DISCUSSION TITLE
Alt-combinations with Kitty on macOS not working (enhanced keyboard protocol question)

### DISCUSSION THREAD

**Discussion by mdomke** [helix-editor/helix#7478] [Q&A]
I have the problem that I cannot use the `Alt` related key combinations like `Alt-.`, `Alt-;`, etc. on macOS using the Kitty terminal emulator (0.28.1). I was reading through the issues
* https://github.com/helix-editor/helix/issues/2469
* https://github.com/helix-editor/helix/issues/4240

where the accepted solution seems to be to use the `macos_option_as_alt = yes` [setting for Kitty](https://sw.kovidgoyal.net/kitty/conf/#opt-kitty.macos_option_as_alt). The problem with that is, that I cannot access the alternate symbols on my keyboard anymore. Let's say I was using the "US International" layout, pressing `Alt-o` should give me the character "ø", but instead I get
```
λ showkey -a 
<ESC>o
```
In my head, the [enhanced keyboard protocol](https://github.com/helix-editor/helix/wiki/Terminal-Support#enhanced-keyboard-protocol) is supposed to help the editor understand which key combination has been pressed and not only provide the resulting text, but maybe I'm wrong.

If I was pressing `Alt-.` Kitty should actually report the following information to Helix if the `macos_option_as_alt` is set to `no`
```
Alt+LEFT_ALT PRESS
CSI 57443 ; 3 u

Alt+. PRESS ≥
CSI 46 ; 3 ; 8805 u

Alt+. RELEASE
CSI 46 ; 3 : 3 u

LEFT_ALT RELEASE
CSI 57443 ; 1 : 3 u
```
so that in normal-mode this could be interpreted as `repeat_last_motion` and in insert-mode this could be the Unicode character "≥". 

### tl:dr
How can I use the alternate keys on my Keyboard and still use the `Alt` key-combinations with Kitty on macOS?

### Version info
```
λ hx --version 
helix 23.05 (7f5940be)
```

```
λ kitty --version
kitty 0.28.1 created by Kovid Goyal
```

**Comment by the-mikedavis** [comment]
The enhanced keyboard protocol allows us to distinguish between terminal codes that are otherwise ambiguous (for example `ret` has the same escape sequence as `C-m`, `tab` as `C-i` and `esc` as `C-[`) and listen for codes that aren't otherwise sent by a host terminal (for example press/release events or the super/meta key). Option's unicode input ability is a native MacOS thing and setting the `macos_option_as_alt` config option tells MacOS to not interpret Option keybinds as unicode input. AFAIK there isn't any interaction between them: we don't get the unicode input text when that option is set to `yes`, just the keycodes.

**Reply by mdomke** [reply]
I do understand that Helix is not able to render the appropriate Unicode character in insert-mode when `macos_option_as_alt=yes` is set, because it is just getting
```
Alt+. PRESS
CSI 46 ; 3 u
```
which translates to `.` (Unicode 46) with modifier `alt` (`1 + 0b10`). In this case, Helix will write just the character `.` into the buffer. In normal-mode however, the `Alt-.` key-press is interpreted correctly as `repeat_last_motion`.

If `macos_option_as_alt=no` is set, Kitty will send (`kitty +kitten show_key -m kitty`)
```
Alt+. PRESS ≥
CSI 46 ; 3 ; 8805 u
```
where additionally the resulting text `≥` (Unicode 8805) is sent as the third parameter. The first and second parameter actually remain the same.

What I don't understand is why Helix cannot interpret the `Alt-.` in normal-mode as `repeat_last_motion` in the last example. It should have all the relevant keycodes available and would just have to ignore the last `text-as-codepoint` parameter. I mean, I haven't looked at the implementation yet, and I'm still giving my opinion from 50 miles away based on assumptions how this [protocol](https://sw.kovidgoyal.net/kitty/keyboard-protocol/#an-overview) could have been implemented. In any way I would be interested if there is a solution to this problem, because it would be great to be able to enter Unicode characters **and** use the full functionality of Helix.

**Reply by the-mikedavis** [reply]
Ah ok yeah, it looks like we can get the unicode text when `macos_option_as_alt=no`. We would need more of the enhanced keyboard protocol though: we would need to enable [report all keys as escape codes](https://sw.kovidgoyal.net/kitty/keyboard-protocol/#report-all-keys-as-escape-codes) and [report associated text](https://sw.kovidgoyal.net/kitty/keyboard-protocol/#report-associated-text) and then use that associated text when inserting. Currently we only use [disambiguate escape codes](https://sw.kovidgoyal.net/kitty/keyboard-protocol/#disambiguate-escape-codes) and [report alternate keys](https://sw.kovidgoyal.net/kitty/keyboard-protocol/#report-alternate-keys). [report associated text](https://sw.kovidgoyal.net/kitty/keyboard-protocol/#report-associated-text) needs to be implemented upstream in [crossterm](https://github.com/crossterm-rs/crossterm) first. I'll make an issue to cover this (edit: https://github.com/helix-editor/helix/issues/7486).

**Reply by mdomke** [reply]
I think I'm misunderstanding how the process of handling key events works exactly. When there is no special treatment from the terminal emulator (e.g. `macos_option_as_alt=no`) macOS will do its `Alt`-keycombo-thing to insert "special" Unicode characters. This works just fine with Helix and everywhere else! This means that in **insert-mode** I can access the full range of characters on my keyboard.
The main problem I see is that applications that want to use an `Alt`-keycombo themselves have trouble finding out if the user wanted to press `Alt-.` when they actually just get `≥`. In my understanding, the enhanced keyboard protocol could help here, because it gives you the information about which keys exactly have been pressed. My assumption was therefore that regardless of the `macos_option_as_alt`-setting Helix could interpret either
```
CSI 46 ; 3 u
```
or
```
CSI 46 ; 3 ; 8805 u
```
as `Alt-.` when in **normal-mode** since `CSI 46 ; 3` means just that. I cannot tell if all of this makes sense in the scope of how the editor handles key press events, but in my understanding, there could be a way for Helix to use the `Alt`-keymappings and allow the macOS users to insert non-ASCII characters.

**Reply by the-mikedavis** [reply]
We don't currently receive `CSI 46 ; 3 ; 8805 u`. The enhanced keyboard protocol is enabled [progressively](https://sw.kovidgoyal.net/kitty/keyboard-protocol/#progressive-enhancement) so we only get the parts of the sequences that we ask for. Currently we only ask for two parts: https://github.com/helix-editor/helix/blob/e0bb032f0e6dcc9b84e571c345928fba4054855e/helix-tui/src/backend/crossterm.rs#L130-L138

`kitty +kitten show_key -m kitty` enables all parts of the protocol. With what we enable now and with `macos_option_as_alt=no`, we receive `≥` (no CSI, just a UTF8 byte sequence) so we don't see any modifiers being pressed or `.`. With `macos_option_as_alt=yes` we only get `A-.` and no mention of `≥`.

It should be possible to interpret `A-.` as both `A-.` for the sake of keybindings and `≥` in insert mode though. We would just need the "report associated text" and probably the "report all keys as escape codes" parts of the protocol implemented upstream and enabled within Helix.

**Reply by mdomke** [reply]
What do you mean with "implemented upstream"? Is this a missing feature in crossterm? It is kinda unfortunate that I can only either type umlauts with my keyboard **or** use the many `Alt`-based key combos in Helix, but not both. I would also try to contribute, but I might need some pointers where to start.

**Reply by the-mikedavis** [reply]
Yep, the "report associated text" part isn't implemented yet in crossterm. See https://github.com/crossterm-rs/crossterm/blob/1efdce7ef63cba6992729db5f22262a60936fa8b/src/event.rs#L266-L269: in the initial PR that added support for the protocol, that part was left for future work. See https://github.com/crossterm-rs/crossterm/pull/754 as an example for expanding support.

**Reply by pr0g** [reply]
I ran into a similar issue to the original question (I wanted `Alt-C` to work on macOS to extend the cursor up - `copy_selection_on_prev_line` using Helix in Kitty).

I was able to get it working by adding the following to my `kitty.conf`:

```
macos_option_as_alt left
map opt+c send_text all \x1bc
```

I hope this helps someone else, it seems to be working well for me now.

**Reply by peterjc** [reply]
See also https://github.com/crossterm-rs/crossterm/issues/965
